pkg/store/prometheus.go

```diff
@@ -220,18 +230,66 @@ func (p *PrometheusStore) Series(r *storepb.SeriesRequest, s storepb.Store_Serie
 	return p.handleStreamedPrometheusResponse(s, httpResp, queryPrometheusSpan, extLset)
 }
 
-func (p *PrometheusStore) handleSampledPrometheusResponse(s storepb.Store_SeriesServer, httpResp *http.Response, querySpan tracing.Span, extLset labels.Labels) error {
-	ctx := s.Context()
+func (p *PrometheusStore) queryPrometheus(s storepb.Store_SeriesServer, r *storepb.SeriesRequest) error {
+	step := math.Max(float64(r.QueryHints.StepMillis/1000), 30)
+	minTime := math.Max(float64(r.MinTime), float64(r.MaxTime-5*time.Minute.Milliseconds()))
+	opts := promclient.QueryOptions{}
+	matrix, _, err := p.client.QueryRange(s.Context(), p.base, r.ToPromQL(), int64(minTime), r.MaxTime, int64(step), opts)
+	if err != nil {
+		return err
+	}
+
+	for _, vector := range matrix {
+		var lbls []labels.Label
```

Comment:
You could potentially make this a bit faster by preallocating:

```go
externalLbls := p.externalLabelsFn()
lbls := make([]labels.Label, 0, len(externalLbls)+len(vector.Metric))
...
```

Reply:
Thanks, updated

Reply:
Looks like we have this report now: https://github.com/thanos-io/thanos/pull/4917/files#annotation_2368016155

Not sure if that's important thought.